hazelcast/src/main/java/com/hazelcast/map/impl/tx/TransactionalMapProxy.java

```diff
@@ -126,29 +126,29 @@ public Object getForUpdate(Object key) {
     }
 
     @Override
-    public Object put(Object key, Object value) {
+    public V put(K key, V value) {
         return put(key, value, -1, MILLISECONDS);
     }
 
     @Override
-    public Object put(Object key, Object value, long ttl, TimeUnit timeUnit) {
+    public V put(K key, V value, long ttl, TimeUnit timeUnit) {
         checkTransactionState();
         MapService service = getService();
         MapServiceContext mapServiceContext = service.getMapServiceContext();
         Data keyData = mapServiceContext.toData(key, partitionStrategy);
-        Object valueBeforeTxn = toObjectIfNeeded(putInternal(keyData, mapServiceContext.toData(value), ttl, timeUnit));
+        V valueBeforeTxn = toObjectIfNeeded(putInternal(keyData, mapServiceContext.toData(value), ttl, timeUnit));
```

Comment:
Hello @LoneRifle. yes, this is a correct observation. 
when the proxy is used on behalf of clients then it might return `Data` instead of deserialized value. It has 2 effects:
1. better performance as it avoids de-serialization and subsequent serialization back to Data (to be sent to a client)
2. the domain class does not need to be available on server classpath.(this is very important)

I think a better solution here is to use a different (internal?) API when executing on behalf of a client. but there might be other solutions as well. We are going to discuss it with @gurbuzali on Monday. @LoneRifle: are you available as well? we could include you in the discussion if you are interested. ;-)